docs/reactivity.md

```diff
@@ -0,0 +1,410 @@
+
+Reactivity also allows incremental evaluation of code when values change: only the code blocks that are “downstream” of changed variables run. This makes interaction and animation more performant because you’re not re-rendering the entire page when state changes.
+
+To be precise, Framework’s reactivity manifests as:
+
+- [Promises](#promises) are implicitly awaited across code blocks
+- [Generators](#generators) are implicitly iterated across code blocks
+- Editing code (or files) triggers reactive updates during preview
+- The [`invalidation` promise](#invalidation) allows clean-up
+
+We’ll cover each of these below.
+
+## Top-level variables
+
+A top-level variable declared in a JavaScript fenced code block can be referenced in another code block or inline expression on the same page. So if you say:
+
+```js echo
+const x = 1, y = 2;
+```
+
+Then you can reference `x` and `y` elsewhere on the page (with values ${x} and ${y}, respectively); top-level variable declarations are effectively [hoisted](https://developer.mozilla.org/en-US/docs/Glossary/Hoisting).
+
+To prevent variables from being visible outside the current block, make them local with a block statement (curly braces):
+
+```js echo
+{
+  const z = 3;
+}
+```
+
+If multiple blocks define top-level variables with the same name, these blocks will still run, but any references to duplicated variables in other blocks will throw a duplicate definition error because the definition is ambiguous.
+
+## Promises
+
+<div class="tip">See <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises">MDN’s <i>Using promises</i></a> for an introduction to promises.</div>
+
+A [promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) represents the result of an asynchronous operation: some data loaded from a file, say, or a module imported dynamically, or the end of an animation, or some text input from the user.
+
+In Framework, when one code block refers to a promise defined in another code block, the referencing code block implicitly awaits the promise. This means promises are often invisible — you don’t have to worry whether something is a promise or whether it’s resolved.
+
+<div class="note">Implicit <code>await</code> only applies <i>across</i> code blocks, not <i>within</i> a code block. Within a code block, a promise is just a promise.</div>
+
+For example, below `FileAttachment.json` returns a promise, and so the value of `volcano` is a promise.
+
+```js echo
+const volcano = FileAttachment("volcano.json").json();
+```
+
+And yet if we reference `volcano` in another code block or inline expression, we don’t need to `await`. The `await` is implicit; the code block automatically waits for the `volcano` promise to resolve before running.
```

Comment:
I think it’s better to stick with promise terminology here. Otherwise it raises questions like does Framework intercept calls to _fetch_ and know whether “data” is “loading”?